Improve code folding to exclude folding line breaks in whitespace-sensitive languages

nigelmj opened the pull request:
Updated the foldable_range method to exclude folding line breaks during code folding in whitespace-sensitive languages like Python and YAML. This adjustment ensures that folding behaves as expected, similar to other code editors.

Ref #11614

Release Notes:

- Fixed #11614

<img width="1219" alt="Screenshot 2024-06-16 at 15 43 31" src="https://github.com/zed-industries/zed/assets/87859239/dd05de16-7f20-4c88-9e95-021555b8b78b">
<img width="1219" alt="Screenshot 2024-06-16 at 15 45 10" src="https://github.com/zed-industries/zed/assets/87859239/b1b78cdd-f34d-4ea3-9728-4741727a9643">

Commit:
Improve code folding to exclude line break folds

Updated the foldable_range method to exclude folding line breaks
during code folding in whitespace-sensitive languages like Python
and YAML. This adjustment ensures that folding behaves as expected,
similar to other code editors.

Ref #11614

SomeoneToIgnore reviewed (commented):
Thank you.

Could we get another test inside `editor_tests.rs` of the same crate, to ensure no future changes regress this?

SomeoneToIgnore commented:
> whitespace-sensitive languages like Python and YAML

Also, given that both the PR title and body mention those explicitly, was this change meant for a limited subset of languages?
Because now this PR changes the folds for all languages at once.

bennetbo commented on crates/editor/src/display_map.rs:
```diff
@@ -984,7 +984,18 @@ impl DisplaySnapshot {
-            Some((start..end, self.fold_placeholder.clone()))
+
+            // Adjust end point to exclude folding line break if present
+            let adjusted_end = if self.buffer_snapshot.is_line_blank(MultiBufferRow(end.row)) {
```
While this works for a single blank line, it does not work for multiple blank lines, e.g. this code:

```python
def py():
    a = 4

    b = 4


def test():
    pass
```

will still miss a line break when it is folded

<img width="197" alt="image" src="https://github.com/zed-industries/zed/assets/53836821/136ae223-7192-46e2-b9fc-b81c89af81b5">

I think this could be solved by just searching upwards until we find a non-empty line, similar too how it's done for indent guides

nigelmj replied:
> While this works for a single blank line, it does not work for multiple blank lines, e.g. this code:
> 
> ```python
> def py():
>     a = 4
> 
>     b = 4
> 
> 
> def test():
>     pass
> ```
> 
> will still miss a line break when it is folded
> 
> <img alt="image" width="197" src="https://private-user-images.githubusercontent.com/53836821/340294634-136ae223-7192-46e2-b9fc-b81c89af81b5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTg2MjU0MjMsIm5iZiI6MTcxODYyNTEyMywicGF0aCI6Ii81MzgzNjgyMS8zNDAyOTQ2MzQtMTM2YWUyMjMtNzE5Mi00NmUyLWI5ZmMtYjgxYzg5YWY4MWI1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjE3VDExNTIwM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTRhZGRlN2I3N2Q0MjNkZjgxNzJlNTJhODM5NzQ3MTI0MzUyOWMxNGM0MTQ3MmUwZjUzNjc3YjZlNGI2NWRmN2ImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.e2QB5-aQnUqD-L7ysHXliLlBqAP1w1RXAZXYd6-zqdY">
> I think this could be solved by just searching upwards until we find a non-empty line, similar too how it's done for indent guides

@bennetbo Thank you for the idea! I will try to get that done as well.

nigelmj commented:
> Thank you.
> 
> Could we get another test inside `editor_tests.rs` of the same crate, to ensure no future changes regress this?

@SomeoneToIgnore Sure, working on that atm!

nigelmj commented:
> > whitespace-sensitive languages like Python and YAML
> 
> Also, given that both the PR title and body mention those explicitly, was this change meant for a limited subset of languages? Because now this PR changes the folds for all languages at once.

Apologies, that was an oversight on my end and I should have made myself clearer. This change was meant for all languages that use whitespace/indentation over curly brackets. I figured that would be better than modifying the behaviour for only a few languages.

Commit:
Test folding for whitespace-sensitive languages

Added a test to verify the folding behaviour in whitespace-sensitive
languages. This test ensures that line breaks are excluded from folding,
thereby maintaining readability of the code.

Ref #11614

nigelmj commented:
@cla-bot check

Commit:
Exclude post-block line breaks in folding

Folding now accounts for multiple line breaks between different
blocks and excludes them in whitespace-sensitive languages.

Ref #11614

Commit:
Test folding for blocks with multiple line breaks

Added a test to verify that the folding behaviour works as expected
for code blocks with multiple line breaks in whitespace-sensitive
languages.

Ref #11614

nigelmj commented:
I have updated the `foldable_range` method to account for multiple line breaks and added new tests to check the fold action behaviour. Folding now works as expected 👍.

SomeoneToIgnore commented on crates/editor/src/display_map.rs:
```diff
@@ -983,8 +983,23 @@ impl DisplaySnapshot {
             }
-            let end = end.unwrap_or(max_point);
-            Some((start..end, self.fold_placeholder.clone()))
+            let mut adjusted_end = end.unwrap_or(max_point);
```
```suggestion
            let mut end_before_line_breaks = end.unwrap_or(max_point);
```

NIT: we could remove both comments in this new code and name the variable a bit more precise.

-----------

Or, we can go a bit further and stop calculating the column in the loop with 
```rust
 self.buffer_snapshot
                        .line_len(MultiBufferRow(adjusted_end.row - 1))
```
and do that only once in the end.

So the variable name above could become `row_before_line_breaks` or similar.

nigelmj replied:
Thank you for the suggestions! I have updated the code accordingly.

SomeoneToIgnore commented on crates/editor/src/display_map.rs:
```diff
@@ -983,8 +983,23 @@ impl DisplaySnapshot {
+                    .buffer_snapshot
+                    .is_line_blank(MultiBufferRow(adjusted_end.row))
+            {
+                // Keep searching upwards until we find a non-blank line
```
One concern I have is whether `start..adjusted_end` is legal, when we go back past `start` (as, in theory, this code can do?)

We're lucky that Rust is fine with such ranges and considers them empty, but I think it's better to return `None` or, if there's a meaning in the `adjusted_end..start` range, use that?

nigelmj replied:
I have changed the condition to stop the loop if the row goes past the start. Instead of a `> 0` check, it now checks `> start.row`. I believe this should resolve the issue, but we could also consider returning None or adjusting the range if needed. Let me know your thoughts!

SomeoneToIgnore reviewed (commented):
Thanks, this looks almost ready, I've left a few comments.
@bennetbo do you have anything to add?

Commit:
Rename variable to improve clarity

Renamed variable to better describe its function.
Additionally, moved column calculation outside the loop
to reduce operations.

Ref #11614

Commit:
Refine fold range calculation

Updated loop condition to prevent going beyond the start row. This
ensures folding behavior is accurate.

Ref #11614

nigelmj commented:
@cla-bot check

nigelmj commented:
@cla-bot check

SomeoneToIgnore reviewed (approved):
I would change `row_before_line_breaks` to be the actual row, but it's not super important.

LGTM to me, but let's wait for another confirmation, just in case.
